pubtools/_pulp/tasks/publish.py

```diff
@@ -0,0 +1,221 @@
+
+    This command will publish the pulp repositories provided in the request or
+    fetched using the filters(url-regex or published-before) or an intersection
+    of input repos and filters.
+    """
+
+    def add_args(self):
+        super(Publish, self).add_args()
+
+        self.parser.add_argument(
+            "--repos", help="list of repos to be published", nargs="+"
+        )
+        self.parser.add_argument(
+            "--clean",
+            help="attempt to remove content not in the repo",
+            type=bool,
+            default=None,
+        )
+        self.parser.add_argument(
+            "--force", help="execute a full repo publish", type=bool, default=None
+        )
+        self.parser.add_argument(
+            "--published-before",
+            help="publish the repos last published before given date e.g. 2019-08-21",
+            type=publish_date,
+            default=None,
+        )
+        self.parser.add_argument(
+            "--url-regex",
+            help="publish repos whose repo url match",
+            default=None,
+            type=re.compile,
+        )
+
+    def run(self):
+        to_await = []
+        LOG.info("run publish")
+
+        # get repos applying filters
+        repos = self.get_repos()
+
+        # publish the repos found
+        publish_fs = self.publish(repos)
+
+        # wait for the publish to complete before
+        # flushing caches.
+        f_sequence(publish_fs).result()
+
+        # flush UD cache
+        to_await.extend(self.flush_ud(repos))
+
+        # flush CDN cache
+        to_await.extend(self.flush_cdn(repos))
+
+        # wait for everthing to finish.
+        for f in to_await:
+            f.result()
+
+        LOG.info("Publishing repositories completed")
+
+    @step("Publish")
+    def publish(self, repos):
+        out = []
+
+        publish_opts = PublishOptions(force=self.args.force, clean=self.args.clean)
+        for repo in repos:
+            LOG.info("Publishing %s", repo.id)
+            f = repo.publish(publish_opts)
+            f = f_map(f, partial(self.log_publish, repo))
+            out.append(f)
+
+        return out
+
+    def log_publish(self, repo, tasks):
+        # logs publish status
+        for task in tasks:
+            LOG.info(
+                "Publishing %s %s",
+                repo.id,
+                "successful" if task.succeeded else "unsuccessful",
+            )
+
+    @step("Flush UD cache")
+    def flush_ud(self, repos):
+        client = self.udcache_client
+        if not client:
+            LOG.info("UD cache flush is not enabled.")
+            return []
+
+        out = []
+        for repo in repos:
+            out.append(client.flush_repo(repo.id))
+            if repo.eng_product_id:
+                out.append(client.flush_product(repo.eng_product_id))
+
+        return out
+
+    @step("Flush CDN cache")
+    def flush_cdn(self, repos):
```

Comment:
This seems to be identical to clear_repo.  Think you could find a way to avoid duplicating the code?

Reply:
yeah. moved it to a different class.